### PR TITLE
Add guide on deploying Lagom to DC/OS via Marathon

### PR DESCRIPTION
**Pull request by longshorej**
This PR adds a guide that explains how you can deploy a Lagom system to DC/OS via Marathon and ConstructR. Specifically it covers how Chirper does it (via https://github.com/lagom/activator-lagom-java-chirper/pull/92), so similar to the strategy taken with the Kubernetes guide.

connected to lightbend/orchestration-team#34

### CODE REVIEW

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (TimMoore) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: It would be helpful to mention which config to use. I followed the instructions at https://github.com/dcos/dcos-vagrant#quick-start, which only creates 1 public and 1 private agent, but it wasn't obvious to me at the time what it was doing or that this wouldn't work.

It looks like https://github.com/dcos/dcos-vagrant/blob/master/docs/deploy.md#multi-agent-cluster is the one you want.

Comment (TimMoore) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: We might also want to note memory requirements. The first time I tried to start up that config (on my 16GB laptop) it caused the system to crash. After reboot and ensuring that I closed other RAM-hungry apps I got it to work, but it is bursting at the seams a bit.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (TimMoore) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Suggest combining this into the previous step, since it only makes sense if you have already built the images locally.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (TimMoore) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Even this configuration might be insufficient. I am seeing that marathon-lb is not being deployed due to insufficient memory on the public agent. On the other hand, the private agents have spare capacity. Perhaps we should supply our own config file.

**Review by TimMoore — COMMENTED** [review]
Thanks for writing this up. I couldn't run through this to completion due to resource constraints. Rather than trying to figure it out myself, I'm wondering if you have a config file for a working DC/OS Vagrant configuration.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (longshorej) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Yikes - I'm a bit out of touch with my 32GB  :blush: -- there I'm using the `VagrantConfig-1m-3a-1p.yaml` config (`cp` it to `VagrantConfig.yaml` to "activate" it)

I'll try this with a 16GB machine tomorrow and update the guide to address any pain points and expand on the DC/OS Vagrant configuration.

Comment (longshorej) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Although I will say the amount of ram it requires in general is painful. Even makes Minikube look light weight :stuck_out_tongue:

**Review by fsat — APPROVED** [review]
@longshorej - approved pending the mention of memory requirement, also combining Step 4 and 5 as suggested by @TimMoore

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (longshorej) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Maybe Lagom is a flexible microservices framework? :wink:

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (longshorej) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: No easy way so far to get this to work on 16GB because of Cassandra -- the Universe packaging of it eats up 16GB (3x5GB plus 1x1GB min) minimum out of the box. I guess I'll try running a single node Docker image in Marathon and offering both choices in the guide.

Comment (TimMoore) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: It's probably not that important to get it running in a 16GB environment... it's enough to just warn people away from trying.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (TimMoore) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: 👍

**Review by TimMoore — APPROVED** [review]

**Comment by longshorej** [comment]
I moved this back to WIP because I'm going to solicit some feedback from Ruth and others like we did for the K8s guide.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (rstento) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: "complement running Lagom applications" is maybe a bit vague? Could you supply some "for instance" benefits I might get from deploying my Lagom system on DC/OS?

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (rstento) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Maybe end the first sentence after [Akka Cluster] and replace "and this has ...." with: When deploying to an orchestrated environment such as DC/OS, you need to make sure that the clusters are bootstrapped correctly.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (rstento) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Maybe instead; Lagom applications use a [Service Locator](https://www.lagomframework.com/documentation/1.3.x/java/ServiceLocator.html) for communication between services.  The locator needs to work together with the Mesos DNS service.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (rstento) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: After reading the rest of the readme, maybe: "provides features such as DNS, routing, and clustering services that Lagom systems use in production."

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (rstento) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: This guide uses the Lagom Chirper example , which is already configured to run on DC/OS. It points out the configuration you would need to make in your Lagom system. And, it provides step-by-step instructions for running the Chirper example on DC/OS.

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (rstento) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: Is Kubernetes a typo here?

File: guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md
Comment (rstento) [inline guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: same comment as before, instead of "complement" maybe "many features that a Lagom system needs to run in production?

Comment (rstento) [reply guides/lagom-dcos-marathon-deploy-microservices/src/main/paradox/index.md]: You could end the sentence with a colon and have the next few sentences be bullets.

**Review by rstento — COMMENTED** [review]
Nice job Jason! I've made several comments, use what you like. 

I just wasn't able to try it out at this time.

**Comment by longshorej** [comment]
Thanks @rstento, much appreciated! I'll incorporate your feedback tomorrow.

**Comment by longshorej** [comment]
Given that this _still_ isn't merged, may make sense to update it to use Akka Management.

**Comment by longshorej** [comment]
This is ready for review/merge/publishing.

**Comment by dwijnand** [comment]
Should we merge or close this?

**Comment by TimMoore** [comment]
Seems obsolete now that Lightbend Orchestration supports Marathon and DC/OS.

**Comment by TimMoore** [comment]
Oh, I take it back… I only read the initial comment saying that it uses ConstructR, but I see now that it was updated since then.

**Comment by TimMoore** [comment]
I have never been able to get a working DC/OS environment on my laptop, so I haven't been able to run through or test any of our DC/OS guides, but if someone can verify that these instructions work, then it would be good to merge and publish this.

**Comment by longshorej** [comment]
FWIW, this was working when I last updated it in March. Memory constraints around DC/OS's version of Cassandra are the biggest hurdle for a single node machine.

**Comment by TimMoore** [comment]
Thanks, @longshorej … I'm happy to just go ahead and merge it (but don't have permission to do so myself) but I still think it would be a good idea for someone else to run through it, just so we have more people familiar with the process (e.g., for support purposes).

Does anyone have a 32GB machine handy? 😄

**Comment by longshorej** [comment]
Ah - reviewing again now -- this was updated like a week before the work on LBO4DCOS was done. Unfortunately, that means it is a bit out of date. It assumes a manually crafted marathon.json file. On the bright side it is written with akka-management in mind, so any updates would just have to take the commands on the LBO docs and port them over here.

**Comment by longshorej** [comment]
Regarding permissions, I think @dbrinegar has access.

**Comment by dbrinegar** [comment]
apparently not?  @alanngai ?

image

**Comment by alanngai** [comment]
@TimMoore has write access now

**Comment by longshorej** [comment]
Cleaning out old PRs of mine. If anyone wants to resurrect or consult this guide, I'll keep it around at the following branch:

https://github.com/longshorej/prod-suite-management-doc/tree/lagom-dcos

**Comment by dwijnand** [comment]
No need to keep the branch, you can checkout pull requests themselves: https://help.github.com/articles/checking-out-pull-requests-locally/.

**Pull request closed without merging**
